### PR TITLE
Introduce variables in profile paths

### PR DESCRIPTION
**Pull request by mgerstner**
This is a first proposal to introduce variables in profile paths. Once we agree on the basic syntax, config file and implementation I'll add some additional commits to document and test the new feature.

### CODE REVIEW

File: src/chkstat.cpp [hunk 1]
```diff
@@ -612,8 +639,83 @@ bool Chkstat::parseProfile(const std::string &path)
...
+    expansions.clear();
+    expansions.push_back("");
+
+    // process each path component.
+    //
+    // we support variables only as individual
+    // path components i.e. something like %{myvar}stuff/suffix is not
+    // allowed, only %{myvar}/suffix.
+    //
+    // multiple variable components in the same path are supported
```
Comment (maltek) [inline src/chkstat.cpp:677]: Implicitly this parsing means we don't support slashes in variable *names*. That's perfectly fine. I'm just wondering if we should make this more explicit.

File: src/chkstat.cpp [hunk 2]
```diff
@@ -612,8 +639,83 @@ bool Chkstat::parseProfile(const std::string &path)
...
+
+            // now we need to create additional entries for each possible
+            // value of the variable
+            std::vector<std::string> new_expansions;
+
+            for (const auto &element: expansions)
+            {
+                for (const auto &var_value: it->second)
+                {
+                    std::string sep = var_value[0] == '/' ? "" : "/";
```
Comment (maltek) [inline src/chkstat.cpp:700]: Can `var_value` ever be empty here? Maybe we can enforce all variable values starting with a slash when reading them in?

**Review by maltek — COMMENTED** [review]
Looking good overall.

The commit message still talks about "${variable}" (with $) in the third paragraph.

Some tests would be nice, including for the edge cases I brought up.

File: src/utility.h [hunk 3]
```diff
@@ -44,6 +44,42 @@ class SwitchArgRW :
...
+ * \brief
+ *  ValueArg with sane const semantics
+ * \details
+ *  The TCLAP::ValueArg is missing accessors that allow access the contained
+ *  value in const contexts. Sadly there was no release of TCLAP in a long
+ *  time, the upstream master branch contains suitable fixes already, however.
+ *  This is just a small wrapper to fix this situation.
+ **/
+template <typename T>
+class SaneValueArg :
```
[the remark this answers is not in the document]
Comment (mgerstner) [reply src/utility.h:57]: Yeah I don't know ... also sounds kind of strange, since it isn't really const, it just supports const usage.

File: src/chkstat.cpp (hunk 1 above)
Comment (mgerstner) [reply src/chkstat.cpp:677]: I think no special characters should be allowed in variable names. alphanumerical only. plus underscore maybe. I can add checking logic for that.

File: src/chkstat.cpp (hunk 2 above)
Comment (mgerstner) [reply src/chkstat.cpp:700]: Since the parser currently splits words at whitespace boundaries for the variable values no value should ever be empty. Even then we would only access the string terminator here.

I thought about requiring them starting with a slash, too. I fear, however, that this could limit the usability of variables or at least make them strange to use if you ever happen to want it to contain only a single path component.

Comment (mgerstner) [reply src/chkstat.cpp:700]: Maybe I should normalize the variables by removing leading slashes and also duplicate ones and trailing ones. How about that?

Comment (maltek) [reply src/chkstat.cpp:700]: Yes, that makes sense.

File: src/chkstat.cpp (hunk 1 above)
Comment (maltek) [reply src/chkstat.cpp:677]: Also makes sense :)

File: src/utility.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (maltek) [reply src/utility.h:57]: True. I don't have any other good ideas either. ('Constable' has a very different meaning.)

[the remark this answers is not in the document]
Comment (mgerstner) [reply src/utility.h:57]: I'd say we leave it as is

**Comment by mgerstner** [comment]
The additional comit contains two additions for limiting supported variable characters and for normalizing variable value paths. I also adjusted the bad commit comment.

**Comment by mgerstner** [comment]
I added commits for the documentation and the test coverage.

**Review by maltek — APPROVED** [review]
LGTM
